kotlinx-coroutines-core/jvm/src/internal/LockFreeLinkedList.kt

```diff
@@ -233,55 +236,71 @@ public actual open class LockFreeLinkedListNode {
      * In particular, invoking [nextNode].[prevNode] might still return this node even though it is "already removed".
      * Invoke [helpRemove] to make sure that remove was completed.
      */
-    public actual open fun remove(): Boolean {
+    public actual open fun remove(): Boolean =
+        removeOrNext() == null
+
+    // returns null if removed successfully or next node if this node is already removed
+    @PublishedApi
+    internal fun removeOrNext(): Node? {
         while (true) { // lock-free loop on next
             val next = this.next
-            if (next is Removed) return false // was already removed -- don't try to help (original thread will take care)
-            if (next === this) return false // was not even added
+            if (next is Removed) return next.ref // was already removed -- don't try to help (original thread will take care)
+            if (next === this) return next // was not even added
             val removed = (next as Node).removed()
             if (_next.compareAndSet(next, removed)) {
                 // was removed successfully (linearized remove) -- fixup the list
-                finishRemove(next)
-                return true
+                next.correctPrev(null)
+                return null
             }
         }
     }
 
+    // Helps with removal of this node
     public actual fun helpRemove() {
-        val removed = this.next as? Removed ?: error("Must be invoked on a removed node")
-        finishRemove(removed.ref)
+        // Note: this node must be already removed
```

Comment:
it should be asserted then?

Reply:
Cast on the next line will fail if that is not the case.